### DISCUSSION TITLE
Use a custom test-runner [kotlin]

### DISCUSSION THREAD

**Discussion by rtc11** [com-lihaoyi/mill#4034] [Q&A] [answered]
How can I add my custom TestRunner to mill?
I tried the following, but I get this error when running `mill test`:
```
test.discoveredTestClasses java.lang.NoSuchMethodException: test.TesterKt.<init>()
    java.base/java.lang.Class.getConstructor0(Class.java:3761)
    java.base/java.lang.Class.getDeclaredConstructor(Class.java:2930)
    mill.testrunner.Framework$.framework(Framework.scala:9)
    mill.scalalib.TestModule.$anonfun$discoveredTestClasses$5(TestModule.scala:50)
    mill.util.Jvm$.inprocess(Jvm.scala:372)
    mill.scalalib.TestModule.$anonfun$discoveredTestClasses$3(TestModule.scala:43)
```

Here is my build.mill:
```scala
package build
import mill._, kotlinlib._

object `package` extends RootModule with KotlinModule {
    def kotlinVersion = "2.0.21"
    
    def sources = T.sources {
      super.sources() ++ Seq(PathRef(millSourcePath / "src"))
    }
    def resources = T.sources {
      super.resources() ++ Seq(PathRef(millSourcePath / ".res"))
    }
    def ivyDeps = Agg(
      ivy"org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.9.0"
    )

    object test extends KotlinTests with TestModule {
        override def testFramework: T[String] = "test.TesterKt"

        def sources = T.sources {
          super.sources() ++ Seq(PathRef(millSourcePath / "test"))
        }
    }
}
```

My project structure looks like this, where the Tester.kt has a main function.
```
.
├── build.mill
├── src
│   ├── server
│   │   └── Server.kt
│   ├── test
│   │   ├── Assertions.kt
│   │   └── Tester.kt
├── test
│   ├── server
│   │   └── ServerTest.kt
```

**Comment by lihaoyi** [comment]
@rtc11 looks like your source configuration is incorrect. Just eyeballing `build.mill`, it configures one main module with sources in `src/` (thus containing both `src/server/` and `src/test/`) and one test module with sources in `test/src` or `test/test/`

You probably want something like

```scala
package build
import mill._, kotlinlib._

object `package` extends RootModule with KotlinModule {
    def kotlinVersion = "2.0.21"
    
    def sources = T.sources {
      super.sources() ++ Seq(PathRef(millSourcePath / "src"))
    }
    def resources = T.sources {
      super.resources() ++ Seq(PathRef(millSourcePath / ".res"))
    }
    def ivyDeps = Agg(
      ivy"org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.9.0"
    )

    object test extends KotlinTests with TestModule {
        override def testFramework: T[String] = "test.TesterKt"

        def sources = T.sources {
          super.sources() ++ Seq(PathRef(millSourcePath / "server"))
        }
    }
}
```

You can debug this by running `./mill show sources` and `./mill show test.sources` to see where those source folder paths have been configured to live

**Reply by rtc11** [reply]
Thanks, those commands were useful. I thought `millSourcesPath` pointed at where the build.mill is located, but this is only the case for the root-module? In the test-module the `millSourcePath` seems to be pointing av `test/`.
I solved by case by using:
```
def sources = T.sources {
    Seq(PathRef(millSourcePath))
}
``` 

Can I create my own TestModule trait? Or am I supposed to override definitions like I try to do here?
Im not familiar with Scala yet.

**Reply by lihaoyi** [reply, marked as answer]
You probably don't need to inherit `TestModule` directly, since it comes as part of `KotlinTests`. 

You can move shared module configuration into `trait`s that you re-use where-ever necessary, as discussed in https://mill-build.org/mill/kotlinlib/intro.html#_multi_module_project. This applies to both normal modules and test modules (and any other modules you create)
